plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeTransactionLogIterator.java

```diff
@@ -0,0 +1,115 @@
+ * limitations under the License.
+ */
+package io.trino.plugin.deltalake.transactionlog;
+
+import io.trino.filesystem.TrinoFileSystem;
+import io.trino.plugin.deltalake.transactionlog.checkpoint.LastCheckpoint;
+import io.trino.spi.TrinoException;
+import org.apache.hadoop.fs.Path;
+
+import java.io.IOException;
+import java.util.Iterator;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.Optional;
+
+import static com.google.common.base.Verify.verify;
+import static io.trino.plugin.deltalake.DeltaLakeErrorCode.DELTA_LAKE_INVALID_SCHEMA;
+import static io.trino.plugin.deltalake.transactionlog.TransactionLogParser.readLastCheckpoint;
+import static io.trino.plugin.deltalake.transactionlog.TransactionLogUtil.getTransactionLogDir;
+import static io.trino.plugin.deltalake.transactionlog.TransactionLogUtil.getTransactionLogJsonEntryPath;
+import static io.trino.plugin.deltalake.transactionlog.checkpoint.TransactionLogTail.getEntriesFromJson;
+import static java.util.Objects.requireNonNull;
+
+public class DeltaLakeTransactionLogIterator
+        implements Iterator<List<DeltaLakeTransactionLogEntry>>
+{
+    private final TrinoFileSystem fileSystem;
+    private final Path transactionLogDir;
+    private long version;
+    private Optional<List<DeltaLakeTransactionLogEntry>> entries = Optional.empty();
+    private final Optional<Long> endVersion;
+    private final Optional<Long> lastCheckpointVersion;
+
+    private boolean stopped;
+
+    public DeltaLakeTransactionLogIterator(
+            TrinoFileSystem fileSystem,
+            Path tableLocation,
+            Optional<Long> startVersion,
+            Optional<Long> endVersion)
+    {
+        verify(startVersion.orElse(0L) <= endVersion.orElse(startVersion.orElse(0L)), "startVersion must be less or equal than endVersion");
+        version = startVersion.orElse(0L) - 1;
+        this.endVersion = endVersion;
+        this.fileSystem = requireNonNull(fileSystem, "fileSystem is null");
+        transactionLogDir = getTransactionLogDir(requireNonNull(tableLocation, "tableLocation is null"));
+        lastCheckpointVersion = readLastCheckpoint(fileSystem, tableLocation).map(LastCheckpoint::getVersion);
+    }
+
+    @Override
+    public boolean hasNext()
+    {
+        if (entries.isPresent()) {
+            return true;
+        }
+        entries = retrieveNextEntries();
+        return entries.isPresent();
+    }
+
+    @Override
+    public List<DeltaLakeTransactionLogEntry> next()
+    {
+        if (entries.isPresent()) {
+            List<DeltaLakeTransactionLogEntry> next = entries.get();
+            entries = Optional.empty();
+            return next;
+        }
+
+        return retrieveNextEntries().orElseThrow(NoSuchElementException::new);
+    }
+
+    private Optional<List<DeltaLakeTransactionLogEntry>> retrieveNextEntries()
+    {
+        if (stopped) {
+            return Optional.empty();
+        }
```

Comment:
`stopped` seems unnecessary to me, but maybe I'm missing something

Reply:
without `stopped` we could call `hasNext()` several times and this would incur further unnecessary reads in case when the `endVersion` is not present. Same thing for the situation when a transaction log in between - e.g. `1,2,3,5,6` - note that `4` is missing - is missing.